packages/core/src/services/documents/destroyOrSoftDeleteDocuments.ts

```diff
@@ -0,0 +1,156 @@
+import { omit } from 'lodash-es'
+
+import { database } from '$core/client'
+import { Result, Transaction, TypedResult } from '$core/lib'
+import { Commit, DocumentVersion, documentVersions } from '$core/schema'
+import { and, eq, inArray, ne } from 'drizzle-orm'
+
+async function findUuidsInOtherCommits({
+  tx,
+  documents,
+  commitId,
+}: {
+  documents: DocumentVersion[]
+  commitId: number
+  tx: typeof database
+}) {
+  const uuids = documents.map((d) => d.documentUuid)
+  const docs = await tx
+    .select()
+    .from(documentVersions)
+    .where(
+      and(
+        inArray(documentVersions.documentUuid, uuids),
+        ne(documentVersions.commitId, commitId),
+      ),
+    )
+
+  return docs.map((d) => d.documentUuid)
+}
+
+function getToBeSoftDeleted({
+  documents,
+  existingUuids,
+}: {
+  documents: DocumentVersion[]
+  existingUuids: string[]
+}) {
+  return documents.filter((d) => existingUuids.includes(d.documentUuid))
+}
+
+/**
+ * Destroy documents that were created in other commits in the past
+ * This is a hard delete to don't let shit in the DB
+ */
+async function hardDestroyDocuments({
+  documents,
+  existingUuids,
+  tx,
+}: {
+  documents: DocumentVersion[]
+  existingUuids: string[]
+  tx: typeof database
+}) {
+  const uuids = documents
+    .filter((d) => !existingUuids.includes(d.documentUuid))
+    .map((d) => d.documentUuid)
+  if (uuids.length === 0) return
+  return tx
+    .delete(documentVersions)
+    .where(inArray(documentVersions.documentUuid, uuids))
+}
+
+async function createDocumentsAsSoftDeleted({
+  toBeSoftDeleted,
+  commitId,
+  tx,
+}: {
+  toBeSoftDeleted: DocumentVersion[]
+  commitId: number
+  tx: typeof database
+}) {
+  const toBeCreated = toBeSoftDeleted.filter((d) => d.commitId !== commitId)
+
+  if (!toBeCreated.length) return
+
+  return tx.insert(documentVersions).values(
+    toBeCreated.map((d) => ({
+      ...omit(d, ['id', 'updatedAt', 'createdAt']),
+      deletedAt: new Date(),
+      commitId,
+    })),
+  )
+}
+
+async function updateDocumetsAsSoftDeleted({
+  toBeSoftDeleted,
+  commitId,
+  tx,
+}: {
+  toBeSoftDeleted: DocumentVersion[]
+  commitId: number
+  tx: typeof database
+}) {
+  const uuids = toBeSoftDeleted
+    .filter((d) => d.commitId === commitId)
+    .map((d) => d.documentUuid)
+
+  if (!uuids.length) return
+
+  return tx
+    .update(documentVersions)
+    .set({ deletedAt: new Date() })
+    .where(
+      and(
+        inArray(documentVersions.documentUuid, uuids),
+        eq(documentVersions.commitId, commitId),
+      ),
+    )
+}
+
+async function invalidateDocumentsCacheInCommit(
+  commitId: number,
+  tx = database,
+) {
+  return tx
+    .update(documentVersions)
+    .set({ resolvedContent: null })
+    .where(eq(documentVersions.commitId, commitId))
+}
+
+/**
```

Comment:
Thanks to Carlos for helping me with this. A lot is going on here. I tried my best to make as understandable as possible and added tests